### ISSUE TITLE
Randoop throws java.lang.StackOverflowError exception when generating tests involving generic types

### ISSUE THREAD

**Issue by hsanchez** [randoop/randoop#107]
@mernst, and @bjkeller 

I am trying to use Randoop to generate tests for the [jReactPhysics3D](https://github.com/kovertopz/jReactPhysics3D) physics engine project.  It seems Randoop is having trouble resolving generic types. It tries to resolve them, but it gets stuck in an infinite loop until a StackOverFlowError is thrown. As a result, only a few tests (involving non generic types) are generated. Here is just sample of the log (See entire content in the attached log file):

```
Analyzing jReactPhysics3D
Running some_repo/tools/do-like-javac/dljc -o dljc-out -t dyntrace,graphtool --graph-jar some_repo/libs/prog2dfg.jar --dyntrace-libs some_repo/libs -- mvn package
Running java -ea -classpath needed_libraries_and_classes:some_repo/libs/randoop.jar randoop.main.Main gentests --classlist=/var/folders/91/xrdc3kfx62bdfwzwbygv36tr0000gp/T/clistjTgoii.txt --timelimit=10 --junit-reflection-allowed=false --silently-ignore-bad-class-names=true --junit-output-dir=dljc-out/test-src1 --outputlimit=20
policy = sun.security.provider.PolicyFile@2d98a335
Ignoring abstract class net.smert.jreactphysics3d.body.Body specified via --classlist or --testclass.
Ignoring abstract class net.smert.jreactphysics3d.collision.broadphase.BroadPhaseAlgorithm specified via --classlist or --testclass.
Ignoring abstract class net.smert.jreactphysics3d.collision.narrowphase.NarrowPhaseAlgorithm specified via --classlist or --testclass.

Throwable thrown while handling command: java.lang.StackOverflowError
java.lang.StackOverflowError
	at randoop.types.ParameterizedType.forType(ParameterizedType.java:75)
	at randoop.types.ClassOrInterfaceBound.forType(ClassOrInterfaceBound.java:34)
	at randoop.types.ParameterBound.forType(ParameterBound.java:68)
	at randoop.types.ParameterBound.forTypes(ParameterBound.java:40)
	at randoop.types.TypeVariable.forType(TypeVariable.java:92)
	at randoop.types.ReferenceType.forType(ReferenceType.java:116)
	at randoop.types.WildcardArgumentWithLowerBound.<init>(WildcardArgumentWithLowerBound.java:20)
	at randoop.types.WildcardArgument.forType(WildcardArgument.java:92)
	at randoop.types.TypeArgument.forType(TypeArgument.java:48)
	at randoop.types.ParameterizedType.forType(ParameterizedType.java:87)
	at randoop.types.ClassOrInterfaceBound.forType(ClassOrInterfaceBound.java:34)
....more of the same
```

I am using our [integration test repo](https://github.com/aas-integration/integration-test) for this. I am also using Randoop 3.0. The corpus folder contains only the jReactPhysics3D project (no sorting programs). I also added the build_command.txt, clean_command.txt, and petablox.properties files. The first one has the command: mvn package, the second one: mvn clean, and the third one: petablox.class.path=target/classes/ and petablox.src.path=src/main/. The project's pom.xml excludes these files from being updated with License-related text:

```
<groupId>com.mycila</groupId>
<artifactId>license-maven-plugin</artifactId>
<version>${licensemaven.version}</version>
    <configuration>
          <excludes>
          <exclude>LICENSE.txt</exclude>
          <exclude>nbactions*.xml</exclude>
          <exclude>nb-configuration.xml</exclude>
          <exclude>pom.xml</exclude>
          <exclude>README.md</exclude>
          <exclude>VERSION</exclude>
          <exclude>examples/**</exclude>
          <exclude>build_command.txt</exclude>
          <exclude>clean_command.txt</exclude>
          <exclude>petablox.properties</exclude>
          <exclude>__java_file_names.txt</exclude>
          <exclude>src/main/resources/**</exclude>
          <exclude>src/test/resources/**</exclude>
          </excludes>
          <header>HEADER.txt</header>
          <keywords>
          <keyword>Copyright (c) 2010-2013 Daniel Chappuis</keyword>
          </keywords>
          <mapping>
               <java>SLASHSTAR_STYLE</java>
          </mapping>
          <useDefaultExcludes>true</useDefaultExcludes>
  </configuration>
```  

The generated --classlist=/var/folders/91/xrdc3kfx62bdfwzwbygv36tr0000gp/T/clistjTgoii.txt contains the following classes:

```
net.smert.jreactphysics3d.body.Body
net.smert.jreactphysics3d.body.CollisionBody
net.smert.jreactphysics3d.body.RigidBody
net.smert.jreactphysics3d.collision.BodyIndexPair
net.smert.jreactphysics3d.collision.BroadPhasePair
net.smert.jreactphysics3d.collision.CollisionDetection
net.smert.jreactphysics3d.collision.broadphase.AABBInt
net.smert.jreactphysics3d.collision.broadphase.BodyPair
net.smert.jreactphysics3d.collision.broadphase.BoxAABB
net.smert.jreactphysics3d.collision.broadphase.BroadPhaseAlgorithm
net.smert.jreactphysics3d.collision.broadphase.EndPoint
net.smert.jreactphysics3d.collision.broadphase.NoBroadPhaseAlgorithm
net.smert.jreactphysics3d.collision.broadphase.PairManager
net.smert.jreactphysics3d.collision.broadphase.SweepAndPruneAlgorithm
net.smert.jreactphysics3d.collision.broadphase.Utils
net.smert.jreactphysics3d.collision.narrowphase.NarrowPhaseAlgorithm
net.smert.jreactphysics3d.collision.narrowphase.SphereVsSphereAlgorithm
net.smert.jreactphysics3d.collision.narrowphase.EPA.EdgeEPA
net.smert.jreactphysics3d.collision.narrowphase.EPA.EPAAlgorithm
net.smert.jreactphysics3d.collision.narrowphase.EPA.TriangleComparison
net.smert.jreactphysics3d.collision.narrowphase.EPA.TriangleEPA
net.smert.jreactphysics3d.collision.narrowphase.EPA.TrianglesStore
net.smert.jreactphysics3d.collision.narrowphase.EPA.Utils
net.smert.jreactphysics3d.collision.narrowphase.GJK.GJKAlgorithm
net.smert.jreactphysics3d.collision.narrowphase.GJK.Simplex
net.smert.jreactphysics3d.collision.shapes.AABB
net.smert.jreactphysics3d.collision.shapes.BoxShape
net.smert.jreactphysics3d.collision.shapes.CapsuleShape
net.smert.jreactphysics3d.collision.shapes.CollisionShape
net.smert.jreactphysics3d.collision.shapes.CollisionShapeType
net.smert.jreactphysics3d.collision.shapes.ConeShape
net.smert.jreactphysics3d.collision.shapes.ConvexMeshShape
net.smert.jreactphysics3d.collision.shapes.CylinderShape
net.smert.jreactphysics3d.collision.shapes.SphereShape
net.smert.jreactphysics3d.configuration.ContactsPositionCorrectionTechnique
net.smert.jreactphysics3d.configuration.Defaults
net.smert.jreactphysics3d.configuration.JointsPositionCorrectionTechnique
net.smert.jreactphysics3d.constraint.BallAndSocketJoint
net.smert.jreactphysics3d.constraint.BallAndSocketJointInfo
net.smert.jreactphysics3d.constraint.ContactPoint
net.smert.jreactphysics3d.constraint.ContactPointInfo
net.smert.jreactphysics3d.constraint.FixedJoint
net.smert.jreactphysics3d.constraint.FixedJointInfo
net.smert.jreactphysics3d.constraint.HingeJoint
net.smert.jreactphysics3d.constraint.HingeJointInfo
net.smert.jreactphysics3d.constraint.Joint
net.smert.jreactphysics3d.constraint.JointInfo
net.smert.jreactphysics3d.constraint.JointListElement
net.smert.jreactphysics3d.constraint.JointType
net.smert.jreactphysics3d.constraint.SliderJoint
net.smert.jreactphysics3d.constraint.SliderJointInfo
net.smert.jreactphysics3d.engine.CollisionWorld
net.smert.jreactphysics3d.engine.ConstraintSolver
net.smert.jreactphysics3d.engine.ConstraintSolverData
net.smert.jreactphysics3d.engine.ContactManifold
net.smert.jreactphysics3d.engine.ContactManifoldListElement
net.smert.jreactphysics3d.engine.ContactManifoldSolver
net.smert.jreactphysics3d.engine.ContactPointSolver
net.smert.jreactphysics3d.engine.ContactSolver
net.smert.jreactphysics3d.engine.DynamicsWorld$1
net.smert.jreactphysics3d.engine.DynamicsWorld
net.smert.jreactphysics3d.engine.EventListener
net.smert.jreactphysics3d.engine.Impulse
net.smert.jreactphysics3d.engine.Island
net.smert.jreactphysics3d.engine.Material
net.smert.jreactphysics3d.engine.OverlappingPair
net.smert.jreactphysics3d.engine.ProfileNode
net.smert.jreactphysics3d.engine.ProfileNodeIterator
net.smert.jreactphysics3d.engine.Profiler
net.smert.jreactphysics3d.engine.ProfileSample
net.smert.jreactphysics3d.engine.Timer
net.smert.jreactphysics3d.mathematics.Mathematics
net.smert.jreactphysics3d.mathematics.Matrix2x2
net.smert.jreactphysics3d.mathematics.Matrix3x3
net.smert.jreactphysics3d.mathematics.Quaternion
net.smert.jreactphysics3d.mathematics.Transform
net.smert.jreactphysics3d.mathematics.Vector2
net.smert.jreactphysics3d.mathematics.Vector3
```
 
Here is the complete log file: 
[06152016-jreactphysics3d-randoop-logs.txt](https://github.com/randoop/randoop/files/317022/06152016-jreactphysics3d-randoop-logs.txt)

**Comment by bjkeller** [comment]
@hsanchez 
Thanks for reporting this.  This is a problem with a wildcard bound of the form `? super U`.

I need your help reducing your information to the steps that allow me to build and find the class files. Something like

1. git clone ....
2. cd ...
3. mvn compile
4. the class files are in the directory ....

would allow me to respond more quickly.

Thanks, 
Ben

**Comment by hsanchez** [comment]
@bjkeller 

Ok. I have published the github branch I used to evaluate Randoop 3.0. In the steps below I will describe how to access that branch, how to run our tools, and reproduce the Randoop error. 
 
## Steps to reproduce the issue:

```bash
› git clone https://github.com/aas-integration/integration-test
› cd integration-test
› git fetch 
› git checkout pen
› ./fetch_dependencies.sh // wait until this step finishes  
› python backend.py > logs.txt
```

`› ./fetch_dependencies.sh` will download the needed dependencies (Randoop 3.0 is included) and the tools that will run the analysis step (here is where Randoop throws the StackOverFlowError)

`› python backend.py` will call `../tools/do-like-javac/do_like_javac/tools/dyntrace.py`, which is the python script that calls randoop (line 65). Our `backend.py` will terminate with the following error:

```bash
› python backend.py > logs.txt
Traceback (most recent call last):
  File "backend.py", line 99, in <module>
    main()
  File "backend.py", line 92, in main
    with open(project_kernel_file, "r") as fi: kf.write(fi.read())
IOError: [Errno 2] No such file or directory: '../corpus/jReactPhysics3D/dljc-out/_classes/kernel.txt'
```  

At this point, go ahead and check the `logs.txt.` On the file, you will get information about the location of the file containing Randoop's classlist, the used settings, and the stack trace of the StackOverflowError exception.

**Comment by bjkeller** [comment]
Thanks for the instructions. Unfortunately, the log file I got does not have the class list, and I don't see any class files in the directory.

**Comment by hsanchez** [comment]
@bjkeller 

Try again. It should work now. For some reason the corpus folder was not to included in the `pen` branch. Now it is.

**Comment by hsanchez** [comment]
In the beginning of the log, you should have something like this 

```
Analyzing JReactPhysics3D
STARTED CLEANING PROJECT
FINISHED CLEANING PROJECT
Running repo/tools/do-like-javac/dljc -o dljc-out -t dyntrace,graphtool --graph-jar repo/libs/prog2dfg.jar --dyntrace-libs repo/libs -- mvn package -Dmaven.test.skip=true
Running java -ea -classpath repo/corpus/JReactPhysics3D/target/classes::repo/corpus/JReactPhysics3D/target/classes:repo/libs/randoop.jar randoop.main.Main gentests --classlist=/var/folders/91/xrdc3kfx62bdfwzwbygv36tr0000gp/T/clisttOqRRK.txt --timelimit=10 --junit-reflection-allowed=false --silently-ignore-bad-class-names=true --junit-output-dir=dljc-out/test-src1 --outputlimit=20
```

Which logs the path to a file containing the list of classes; e.g., `classlist=/var/folders/91/xrdc3kfx62bdfwzwbygv36tr0000gp/T/clisttOqRRK.txt`

**Comment by bjkeller** [comment]
Thanks. I was able to get to the problem by going straight to the code repo.

**Comment by bjkeller** [comment]
Issue is actually that recursively defined lower bounds are not being recognized. There is a symmetric case for upper bounds, so should have similar solution.

This case happen b/c encounters `java.lang.Comparable<? super U>` as the bound of `U` in `java.util.function.Function<? super T, ? extends U>`.

**Comment by hsanchez** [comment]
@bjkeller 

Will this be fixed in v.3.0.2? If it will, can you give me an approximate date of when this version will be released? Thank you for all your help on this issue.

**Comment by bjkeller** [comment]
Yes. We usually try to resolve known bugs immediately and make a new release.  Since, I'm taking some time off, I won't be able to work on it more now.  But, I expect this will be fixed and the new release made sometime before 7/8.

**Comment by bjkeller** [comment]
Fixed by ed1e590a238b3ee110b7296b9b9aa104493d92a0
